### PR TITLE
Feature/use embeddings

### PR DESCRIPTION
**Pull request by JeroenMandersloot**
* Refactored `run.py` and related files so that the main pipeline works with the new data structures.
* Added `get_embedding()` to the `FaceImage` dataclass, which allows you to get an image's embedding based on a specified `Architecture`. 
  - This method also supports optionally caching results to disk in addition to caching to memory. Caching to disk is helpful for reducing the time it takes to compute the embeddings the first time in a program's lifetime. After the embedding has been computed once within the same program, the memory caching mechanism takes over regardless of whether on-disk caching is used or not.
* Added a `split_pairs()` method to `data.py` which can split a list of `FacePair` instances into two separate lists where it is guaranteed that no paired identities occur in both sets. This can later be made more generic to split into any number of splits.
* Updated unit tests for the newly introduced methods.

### CODE REVIEW

File: lr_face/data.py [hunk 1]
```diff
@@ -60,16 +63,60 @@ def get_image(self,
         return res
 
     @cache
-    def get_embedding(self, architecture: Architecture) -> np.ndarray:
+    def get_embedding(self,
+                      architecture: Architecture,
```
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply lr_face/data.py:67]: Klopt ja, dit had ik voor nu even losgetrokken omdat dit PR nog geen versioning etc had. Is inmiddels al opgelost in https://github.com/HolmesNL/forensische-gezichtsvergelijking/pull/26

File: lr_face/data.py [hunk 2]
```diff
@@ -630,3 +683,33 @@ def to_array(data: Union[Dataset,
     # If we haven't returned something by now it means an invalid data type
     # was passed along, so we let the user know about that.
     raise ValueError(f'Invalid data type: {type(data)}')
+
+
+def split_pairs(
+        pairs: List[FacePair],
+        fraction_test: float,
+        random_state: Optional[int] = 42
```
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply lr_face/data.py]: Random seed is er uit

File: lr_face/data.py [hunk 3]
```diff
@@ -60,16 +63,60 @@ def get_image(self,
...
+        Internal method that actually computes the embedding based on the
+        specified `architecture` without worrying about caching, etc. Returns a
+        1D array of shape `(embedding_size)`.
+
+        :param architecture: Architecture
+        :return: np.ndarray
+        """
+        image = self.get_image(architecture.resolution, normalize=True)
+        x = np.expand_dims(image, axis=0)
+        embedding_model = architecture.get_embedding_model()
```
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply lr_face/data.py:118]: Is ook al gefixed in https://github.com/HolmesNL/forensische-gezichtsvergelijking/pull/26

File: lr_face/models.py [hunk 4]
```diff
@@ -21,51 +23,40 @@ def fit(self, X, y):
...
     def __str__(self):
         return 'Dummy'
 
 
 class ScorerModel:
-    def __init__(self, embedding_model: tf.keras.Model, name: str):
-        self.embedding_model = embedding_model
-        self.name = name
-        self.cache = dict()
+    def __init__(self, architecture: Architecture):
```
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply lr_face/models.py:34]: Ook gefixed in https://github.com/HolmesNL/forensische-gezichtsvergelijking/pull/26

File: tests/test_utils.py [hunk 5]
```diff
@@ -51,3 +54,77 @@ def func(a, b):
...
+
+    # The third also has a `bias` of 2, so its hash is the same as `instance2`.
+    instance3 = CacheTest(2)
+
+    # When we call `multiply` on the first instance we expect the counter to go
+    # from 0 to 1 for `instance1`, but remain 0 for `instance2`.
+    assert instance1.counter[(2, 3)] == 0
+    assert instance2.counter[(2, 3)] == 0
+    instance1.multiply(2, 3)
+    assert instance1.multiply.cache_info().hits == 0
```
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply tests/test_utils.py:98]: Beetje verwarrende terminologie misschien, maar het aantal cache hits (waar deze regel over gaat) is het aantal keer dat je iets uit de cache hebt gehaald. Deze is dus 0 omdat je `multiply` hiervoor pas 1x hebt aangeroepen, dus er was toen nog geen cache om iets uit op te halen.

File: tests/test_utils.py [hunk 6]
```diff
@@ -51,3 +54,77 @@ def func(a, b):
...
+
+    # When we call `multiply` a second time, we should see no further changes
+    # in the counter.
+    instance1.multiply(2, 3)
+    assert instance1.counter[(2, 3)] == 1
+    assert instance2.counter[(2, 3)] == 0
+
+    # We should see a cache hit. The cache hits are apparently stored at the
+    # class-level, so all other instances of the same class will always have
+    # the same number of cache hits. This behavior is weird and so we should
```
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply tests/test_utils.py:110]: Nu ik er langer over nadenk is het heel logisch met hoe python werkt. Daar is bij elke methode op een class `self` een argument, dus je kunt deze dingen in principe op het class niveau opslaan, want doordat je cache gekoppeld is aan je argumenten, is het indirect alsnog gekoppeld aan je specifieke instance. Je moet alleen niet het aantal hits gaan gebruiken voor logica in je code, omdat die dus het aantal hits voor de hele class telt.

File: lr_face/data.py [hunk 7]
```diff
@@ -630,3 +683,33 @@ def to_array(data: Union[Dataset,
...
+    :param pairs: List[FacePair]
+    :param fraction_test: float
+    :param random_state: Optional[int]
+    :return: Tuple[List[FacePair], List[FacePair]]
+    """
+
+    gss = GroupShuffleSplit(n_splits=1,
+                            test_size=fraction_test,
+                            random_state=random_state)
+    groups = ['|'.join(sorted(x.identity for x in pair)) for pair in pairs]
```
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply lr_face/data.py:713]: Ah okee, ik snap het. Dan had ik het hiervoor verkeerd begrepen, ik dacht dat de manier waarop ik het nu geimplementeerd had al overeenkwam met hoe het in de pipeline zat -- maar ik herinner me nu dat je dit probleem ook over de telefoon had uitgelegd en dat je daarom het splitsen al op image-niveau en niet op pair-niveau doet.

Ik zal dit aanpassen zodat de splitsing op image-niveau gebeurt en dat er geen overlap is tussen de inviduele identiteiten in de calibratie- en testset. Dat lijkt mij ook het meest logisch ja.
